File: Lua/cocktail_shaker_sort.lua
function cocktail_shaker_sort(arr)
    local swapped = true
    local start = 1
    local finish = #arr

    while swapped do
        swapped = false
        for i = start, finish - 1 do
            if arr[i] > arr[i + 1] then
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = true
            end
        end
        if not swapped then break end

        swapped = false
        finish = finish - 1

        for i = finish, start + 1, -1 do
            if arr[i] < arr[i - 1] then
                arr[i], arr[i - 1] = arr[i - 1], arr[i]
                swapped = true
            end
        end
        start = start + 1
    end
end

local arr = {5, 1, 4, 2, 8}
cocktail_shaker_sort(arr)
print(table.concat(arr, " "))

File: Python/cocktail_shaker_sort.py
def cocktail_shaker_sort(arr):
    n = len(arr)
    swapped = True
    start = 0
    end = n - 1

    while swapped:
        swapped = False

        for i in range(start, end):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True

        if not swapped:
            break  

        swapped = False
        end -= 1  

        for i in range(end - 1, start - 1, -1):
            if arr[i] > arr[i + 1]:
                arr[i], arr[i + 1] = arr[i + 1], arr[i]
                swapped = True

        start += 1  

    return arr

arr=[3, 7, 4, 8, 6, 2, 1, 5, 0, -5]
print(cocktail_shaker_sort(arr))